Pester v5

tigerfansga opened the pull request:
#### Description
Update pester test for Pester v5 support

#### Issues Fixed
- Fixes #194 

#### References

#### Checklist
- [x] You actually ran the code that you just wrote, especially if you did just "one last quick change".
- [ ] ~~Comment-based help added/updated, including examples.~~
- [x] [Static analysis](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#static-analysis) is reporting back clean.
- [x] New/changed code adheres to our [coding guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#coding-guidelines).
- [ ] ~~[Formatters were created](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#formatters) for any new types being added.~~
- [ ] ~~New/changed code continues to [support the pipeline](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#pipeline-support).~~
- [ ] ~~Changes to the manifest file follow the [manifest guidance](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#module-manifest).~~
- [x] Unit tests were added/updated and are all passing. See [testing guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#testing).  This includes making sure that all pipeline input variations have been covered.
- [ ] ~~Relevant usage examples have been added/updated in [USAGE.md](https://github.com/microsoft/PowerShellForGitHub/blob/master/USAGE.md).~~
- [x] If desired, ensure your name is added to our [Contributors list](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#contributors)

tigerfansga committed:
Update GitHubUsers.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubAnalytics.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubAssignees.tests.ps1 for PesterV5

tigerfansga committed:
Update GitHubBranches.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubContents.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubCore.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubEvents.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubGistComments.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubGists.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubGraphQl.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubIssueComments.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubIssues.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubLabels.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubMilestones.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubOrganizations.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubProjectCards.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubProjectColumns.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubProjects.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubPullRequests.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubReactions.tests.ps1 for Pester v5

tigerfansga committed:
Remove repository code of conduct tests from GitHubMiscellaneous.tests.ps1

tigerfansga committed:
Update GitHubReleases.tests.ps1 to account for pre-releases

tigerfansga committed:
Update GitHubReleases.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubRepositories.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubRepositoryForks.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubRepositoryTraffic.tests.ps1 for Pester v5

tigerfansga committed:
Update GitHubTeams.tests.ps1 for Pester v5

tigerfansga committed:
Update file formating

tigerfansga committed:
Merge branch 'master' into PesterV5

tigerfansga committed:
Update CONTRIBUTING.md

tigerfansga committed:
Update run-unitTests.yaml to install Pester v5.3.3

HowardWolosky commented:
Thanks @tigerfansga  -- I have this queued for review this week.

tigerfansga committed:
Update CONTRIBUTING.md for Pester v5

HowardWolosky commented:
> Thanks @tigerfansga -- I have this queued for review this week.

Got a bit delayed -- I have started reviewing this, but it's going to take me a couple more days though to finish getting through it.  Thanks for your contribution (and your patience)!

tigerfansga commented:
> > Thanks @tigerfansga -- I have this queued for review this week.
> 
> 
> 
> Got a bit delayed -- I have started reviewing this, but it's going to take me a couple more days though to finish getting through it.  Thanks for your contribution (and your patience)!

No problem. It's a lot to go through.

HowardWolosky commented on Tests/GitHubRepositories.tests.ps1:
```diff
@@ -829,970 +741,1060 @@ try
 
-                    $permissions = Get-GitHubRepositoryActionsPermission -Uri $repo.svn_url
-                }
+        $allowedActions = 'All', 'LocalOnly', 'Selected', 'Disabled'
```
```suggestion
```

Looks like we either drop that, or reference it at 1391 (I think dropping it is fine).

HowardWolosky commented on Tests/GitHubRepositories.tests.ps1:
```diff
@@ -829,970 +741,1060 @@ try
+    BeforeAll {
+        $repo = New-GitHubRepository -RepositoryName ([Guid]::NewGuid().Guid)
+
+        $allowedActions = 'All', 'LocalOnly', 'Selected', 'Disabled'
```
```suggestion
```

Looks like we either drop that, or reference it at 1446 (I think dropping it is fine).

HowardWolosky commented on Tests/GitHubRepositories.tests.ps1:
```diff
@@ -829,970 +741,1060 @@ try
-
-                    $repoPermission = Get-GitHubRepositoryTeamPermission @getGithubRepositoryTeamPermissionParms
-                }
+        $permissions = 'Push', 'Pull', 'Maintain', 'Triage', 'Admin'
```
```suggestion
```

Looks like we either drop that, or reference it at 1510 (I think dropping it is fine).

HowardWolosky commented on Tests/GitHubRepositories.tests.ps1:
```diff
@@ -829,970 +741,1060 @@ try
+        $team = New-GitHubTeam @newGithubTeamParms
 
-                    $repoPermission = Get-GitHubRepositoryTeamPermission @getGithubRepositoryTeamPermissionParms
+        $permissions = 'Push', 'Pull', 'Maintain', 'Triage', 'Admin'
```
```suggestion
```

Looks like we either drop that, or reference it at 1655 (I think dropping it is fine).

HowardWolosky reviewed (changes requested):
Wow -- thanks SO MUCH for doing this @tigerfansga.  This migration was definitely a labor of love.

My sincere apologies for the delay in getting this review done.  I took on a new role for my day job, and I had to focus on that in the short term.  Now I'm trying to get the PR backlog cleared out before the holidays.

I'm excited about being able to run tests directly from VS Code now as a result of this change.

There are only some minor changes/things that I'm calling out here.  I've mostly skipped some of the whitespace feedback as I really want to get this one in as soon as possible.  The main thing that I noticed is that the Projects tests are broken, but that has nothing to do with you, and everything to do with GitHub migrating to a completely new Projects feature (with a  GraphQL-backed API).  The old Projects that this module currently has support for has been deprecated by GitHub, and new "classic" Projects can no longer be created (which causes all of these tests to fail).  I'll deal with disabling these tests after we get your change merged in.

I'm still needing to do some additional verification of the changes across different platforms, but here's all of the feedback from code inspection.

HowardWolosky committed:
Merge branch 'master' into PesterV5

HowardWolosky commented:
/azp run PowerShellForGitHub-CI

HowardWolosky commented:
/azp run PowerShellForGitHub-CI

HowardWolosky committed:
Remove trailing spaces

HowardWolosky commented:
/azp run PowerShellForGitHub-CI

HowardWolosky commented:
@tigerfansga - You'll need to sign the CLA via the bot using the command `@microsoft-github-policy-service agree`.

Still looking through the test failures to see if there's anything there related to your changes...

tigerfansga commented:
> @tigerfansga - You'll need to sign the CLA via the bot using the command `@microsoft-github-policy-service agree`.
> 
> 
> 
> Still looking through the test failures to see if there's anything there related to your changes...

That is strange on the CLA because I had accepted it. I can get that done later today. I'm not sure I'll be able to review any requested changes till after the New Year.

HowardWolosky committed:
Update GitHubBranches.tests.ps1

Error message has changed.  Modifying test to simply capture that a failure occurred (without validating the string itself)

tigerfansga commented:
@microsoft-github-policy-service agree

HowardWolosky committed:
Merge branch 'master' into PesterV5

HowardWolosky committed:
Update GitHubRepositories.tests.ps1

Removes the specific error message check from the `Throw` validation of some tests since the wording of the exception message keeps changing.

HowardWolosky commented:
/azp run PowerShellForGitHub-CI

HowardWolosky committed:
Merge branch 'master' into PesterV5

HowardWolosky committed:
Update settings.json

Update the PowerShell extension in VS Code to take advantage of the fact that the tests are now Pester 5 compatible.  This allows each individual block to have a Run/Debug link associated with it.

HowardWolosky committed:
Apply suggestions from code review

HowardWolosky committed:
Apply suggestions from code review

HowardWolosky committed:
Update GitHubProjects.tests.ps1

Disabling the tests since they no longer work.

HowardWolosky committed:
Update GitHubProjectColumns.tests.ps1

Disabling since classic projects have been deprecated.

HowardWolosky committed:
Update GitHubProjectCards.tests.ps1

Disabling since classic projects are no longer supported.

HowardWolosky committed:
Update run-unitTests.yaml

HowardWolosky commented on build/pipelines/templates/run-unitTests.yaml:
```diff
@@ -46,8 +46,23 @@ steps:
       script: |
-        $null = New-Item -Path ..\ -Name Pester -ItemType Directory -Force
-        Invoke-Pester -CodeCoverage .\*.ps*1 -CodeCoverageOutputFile ../Pester/coverage.xml -CodeCoverageOutputFileFormat JaCoCo -EnableExit -Strict -OutputFile ../Pester/test-results.xml -OutputFormat NUnitXml
+        # Import the module as otherwise GitHubCore.tests.ps1 has problems since it's testing things in moduls scope.
```
```suggestion
        # Import the module, otherwise GitHubCore.tests.ps1 has problems since it's testing things in module scope.
```

HowardWolosky committed:
Update build/pipelines/templates/run-unitTests.yaml

HowardWolosky committed:
Update GitHubContents.tests.ps1

Removing duplication

HowardWolosky committed:
Update GitHubIssueComments.tests.ps1

HowardWolosky commented:
/azp run PowerShellForGitHub-CI

HowardWolosky committed:
Update run-unitTests.yaml

HowardWolosky commented:
/azp run PowerShellForGitHub-CI

HowardWolosky commented:
I cleaned up the remaining work in here and got it merged in.  Thanks **so much** to @tigerfansga for taking this work on and moving the project forward.  It's HUGE that we can now run individual tests directly within VS Code now that we're on Pester 5.  
Thanks so much for the contribution!
